File: frontend/src/app/core/reportes/reportes.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';

import { environment } from '../../../environments/environment';
import {
  CajaOcupacionOut,
  ClienteRankingOut,
  ConsultaIaOut,
  EnvioEstadoOut,
  FiltroReportes,
  IndicadoresOut,
  MensajeReporteIn,
  ProductoRankingOut,
  ProductoSinMovimientoOut,
  RecepcionPendienteProveedorOut,
  ReservaEstadoOut,
  StockSucursalOut,
  VendedorOut,
  VentaDiariaOut,
  VentaPorSucursalOut,
} from './reportes.models';

function paramsDesdeFiltro(filtro: FiltroReportes, extra: Record<string, string | number> = {}): HttpParams {
  let params = new HttpParams();
  if (filtro.desde) params = params.set('desde', filtro.desde);
  if (filtro.hasta) params = params.set('hasta', filtro.hasta);
  if (filtro.sucursal_id) params = params.set('sucursal_id', filtro.sucursal_id);
  if (filtro.categoria_id) params = params.set('categoria_id', filtro.categoria_id);
  if (filtro.vendedor_id) params = params.set('vendedor_id', filtro.vendedor_id);
  if (filtro.canal) params = params.set('canal', filtro.canal);
  if (filtro.entrega) params = params.set('entrega', filtro.entrega);
  for (const [clave, valor] of Object.entries(extra)) {
    params = params.set(clave, valor);
  }
  return params;
}

@Injectable({ providedIn: 'root' })
export class ReportesService {
  private readonly http = inject(HttpClient);
  private readonly base = `${environment.apiUrl}/reportes`;

  indicadores(filtro: FiltroReportes): Observable<IndicadoresOut> {
    return this.http.get<IndicadoresOut>(`${this.base}/indicadores`, { params: paramsDesdeFiltro(filtro) });
  }

  ventasDiarias(filtro: FiltroReportes): Observable<VentaDiariaOut[]> {
    return this.http.get<VentaDiariaOut[]>(`${this.base}/ventas-diarias`, { params: paramsDesdeFiltro(filtro) });
  }

  ventasPorSucursal(filtro: FiltroReportes): Observable<VentaPorSucursalOut[]> {
    return this.http.get<VentaPorSucursalOut[]>(`${this.base}/ventas-por-sucursal`, {
      params: paramsDesdeFiltro(filtro),
    });
  }

  topProductos(filtro: FiltroReportes, limite = 10): Observable<ProductoRankingOut[]> {
    return this.http.get<ProductoRankingOut[]>(`${this.base}/top-productos`, {
      params: paramsDesdeFiltro(filtro, { limite }),
    });
  }

  stockPorSucursal(sucursalId?: string | null): Observable<StockSucursalOut[]> {
    let params = new HttpParams();
    if (sucursalId) params = params.set('sucursal_id', sucursalId);
    return this.http.get<StockSucursalOut[]>(`${this.base}/stock-por-sucursal`, { params });
  }

  reservasPorEstado(sucursalId?: string | null): Observable<ReservaEstadoOut[]> {
    let params = new HttpParams();
    if (sucursalId) params = params.set('sucursal_id', sucursalId);
    return this.http.get<ReservaEstadoOut[]>(`${this.base}/reservas-por-estado`, { params });
  }

  enviosPorEstado(sucursalId?: string | null): Observable<EnvioEstadoOut[]> {
    let params = new HttpParams();
    if (sucursalId) params = params.set('sucursal_id', sucursalId);
    return this.http.get<EnvioEstadoOut[]>(`${this.base}/envios-por-estado`, { params });
  }

  productosSinMovimiento(sucursalId?: string | null, limite = 50): Observable<ProductoSinMovimientoOut[]> {
    let params = new HttpParams().set('limite', limite);
    if (sucursalId) params = params.set('sucursal_id', sucursalId);
    return this.http.get<ProductoSinMovimientoOut[]>(`${this.base}/productos-sin-movimiento`, { params });
  }

  topClientes(sucursalId?: string | null, limite = 10): Observable<ClienteRankingOut[]> {
    let params = new HttpParams().set('limite', limite);
    if (sucursalId) params = params.set('sucursal_id', sucursalId);
    return this.http.get<ClienteRankingOut[]>(`${this.base}/top-clientes`, { params });
  }

  ocupacionCajas(sucursalId?: string | null): Observable<CajaOcupacionOut[]> {
    let params = new HttpParams();
    if (sucursalId) params = params.set('sucursal_id', sucursalId);
    return this.http.get<CajaOcupacionOut[]>(`${this.base}/ocupacion-cajas`, { params });
  }

  recepcionesPendientes(sucursalId?: string | null): Observable<RecepcionPendienteProveedorOut[]> {
    let params = new HttpParams();
    if (sucursalId) params = params.set('sucursal_id', sucursalId);
    return this.http.get<RecepcionPendienteProveedorOut[]>(`${this.base}/recepciones-pendientes`, { params });
  }

  vendedores(sucursalId?: string | null): Observable<VendedorOut[]> {
    let params = new HttpParams();
    if (sucursalId) params = params.set('sucursal_id', sucursalId);
    return this.http.get<VendedorOut[]>(`${this.base}/vendedores`, { params });
  }

  consultaIa(mensajes: MensajeReporteIn[]): Observable<ConsultaIaOut> {
    return this.http.post<ConsultaIaOut>(`${this.base}/consulta-ia`, { mensajes });
  }
}

File: movil/lib/core/reportes/reportes_service.dart
import '../api.dart';
import 'reportes_models.dart';

/// CU15 - Consultar Reportes e Indicadores, espejo movil de
/// `frontend/src/app/core/reportes/reportes.service.ts`. Un ENCARGADO solo ve su propia
/// sucursal (el backend lo resuelve solo); un ADMIN puede pedir cualquiera o la cadena
/// completa (sucursalId nulo). Sin capa de repositorio: cada metodo pega directo al
/// endpoint que le corresponde.
class ReportesService {
  String _soloFecha(DateTime fecha) =>
      '${fecha.year.toString().padLeft(4, '0')}-'
      '${fecha.month.toString().padLeft(2, '0')}-'
      '${fecha.day.toString().padLeft(2, '0')}';

  // --- Dinamicos: aceptan fecha/sucursal/canal/categoria/entrega ------------------------

  Future<IndicadoresOut> obtenerIndicadores({
    DateTime? desde,
    DateTime? hasta,
    String? sucursalId,
    String? categoriaId,
    String? canal,
    String? entrega,
  }) async {
    final respuesta = await api.get('/reportes/indicadores', query: {
      'desde': desde == null ? null : _soloFecha(desde),
      'hasta': hasta == null ? null : _soloFecha(hasta),
      'sucursal_id': sucursalId,
      'categoria_id': categoriaId,
      'canal': canal,
      'entrega': entrega,
    });
    return IndicadoresOut.desdeJson(respuesta as Map<String, dynamic>);
  }

  Future<List<VentaDiariaOut>> listarVentasDiarias({
    DateTime? desde,
    DateTime? hasta,
    String? sucursalId,
    String? categoriaId,
    String? canal,
    String? entrega,
  }) async {
    final respuesta = await api.get('/reportes/ventas-diarias', query: {
      'desde': desde == null ? null : _soloFecha(desde),
      'hasta': hasta == null ? null : _soloFecha(hasta),
      'sucursal_id': sucursalId,
      'categoria_id': categoriaId,
      'canal': canal,
      'entrega': entrega,
    });
    return comoLista(respuesta).map(VentaDiariaOut.desdeJson).toList();
  }

  Future<List<VentaPorSucursalOut>> listarVentasPorSucursal({
    DateTime? desde,
    DateTime? hasta,
    String? sucursalId,
    String? categoriaId,
    String? canal,
    String? entrega,
  }) async {
    final respuesta = await api.get('/reportes/ventas-por-sucursal', query: {
      'desde': desde == null ? null : _soloFecha(desde),
      'hasta': hasta == null ? null : _soloFecha(hasta),
      'sucursal_id': sucursalId,
      'categoria_id': categoriaId,
      'canal': canal,
      'entrega': entrega,
    });
    return comoLista(respuesta).map(VentaPorSucursalOut.desdeJson).toList();
  }

  Future<List<ProductoRankingOut>> listarTopProductos({
    DateTime? desde,
    DateTime? hasta,
    String? sucursalId,
    String? categoriaId,
    String? canal,
    String? entrega,
    int limite = 10,
  }) async {
    final respuesta = await api.get('/reportes/top-productos', query: {
      'desde': desde == null ? null : _soloFecha(desde),
      'hasta': hasta == null ? null : _soloFecha(hasta),
      'sucursal_id': sucursalId,
      'categoria_id': categoriaId,
      'canal': canal,
      'entrega': entrega,
      'limite': limite,
    });
    return comoLista(respuesta).map(ProductoRankingOut.desdeJson).toList();
  }

  // --- Estaticos: foto del estado actual, sin filtro de fecha ---------------------------

  Future<List<StockSucursalOut>> listarStockPorSucursal({String? sucursalId}) async {
    final respuesta =
        await api.get('/reportes/stock-por-sucursal', query: {'sucursal_id': sucursalId});
    return comoLista(respuesta).map(StockSucursalOut.desdeJson).toList();
  }

  Future<List<ReservaEstadoOut>> listarReservasPorEstado({String? sucursalId}) async {
    final respuesta =
        await api.get('/reportes/reservas-por-estado', query: {'sucursal_id': sucursalId});
    return comoLista(respuesta).map(ReservaEstadoOut.desdeJson).toList();
  }

  Future<List<EnvioEstadoOut>> listarEnviosPorEstado({String? sucursalId}) async {
    final respuesta =
        await api.get('/reportes/envios-por-estado', query: {'sucursal_id': sucursalId});
    return comoLista(respuesta).map(EnvioEstadoOut.desdeJson).toList();
  }

  Future<List<ProductoSinMovimientoOut>> listarProductosSinMovimiento({
    String? sucursalId,
    int limite = 50,
  }) async {
    final respuesta = await api.get(
      '/reportes/productos-sin-movimiento',
      query: {'sucursal_id': sucursalId, 'limite': limite},
    );
    return comoLista(respuesta).map(ProductoSinMovimientoOut.desdeJson).toList();
  }

  Future<List<ClienteRankingOut>> listarTopClientes({
    String? sucursalId,
    int limite = 10,
  }) async {
    final respuesta = await api.get(
      '/reportes/top-clientes',
      query: {'sucursal_id': sucursalId, 'limite': limite},
    );
    return comoLista(respuesta).map(ClienteRankingOut.desdeJson).toList();
  }

  Future<List<CajaOcupacionOut>> listarOcupacionCajas({String? sucursalId}) async {
    final respuesta =
        await api.get('/reportes/ocupacion-cajas', query: {'sucursal_id': sucursalId});
    return comoLista(respuesta).map(CajaOcupacionOut.desdeJson).toList();
  }

  Future<List<RecepcionPendienteProveedorOut>> listarRecepcionesPendientes({
    String? sucursalId,
  }) async {
    final respuesta =
        await api.get('/reportes/recepciones-pendientes', query: {'sucursal_id': sucursalId});
    return comoLista(respuesta).map(RecepcionPendienteProveedorOut.desdeJson).toList();
  }

  /// PENDIENTES 2.19.5: consolidado de existencias, paginado.
  Future<ExistenciasPaginadoOut> listarExistencias({
    String? sucursalId,
    String? categoriaId,
    String? busqueda,
    String? situacion,
    int pagina = 1,
    int tamanioPagina = 50,
  }) async {
    final respuesta = await api.get('/reportes/existencias', query: {
      'sucursal_id': sucursalId,
      'categoria_id': categoriaId,
      'busqueda': (busqueda?.isEmpty ?? true) ? null : busqueda,
      'situacion': situacion,
      'pagina': pagina,
      'tamanio_pagina': tamanioPagina,
    });
    return ExistenciasPaginadoOut.desdeJson(respuesta as Map<String, dynamic>);
  }

  // --- IA / VOZ: mismo patron de tool-use que CU18, sin persistencia --------------------

  Future<ConsultaIaOut> consultarIa(List<MensajeReporte> mensajes) async {
    final respuesta = await api.post(
      '/reportes/consulta-ia',
      cuerpo: {'mensajes': mensajes.map((m) => m.aJson()).toList()},
    );
    return ConsultaIaOut.desdeJson(respuesta as Map<String, dynamic>);
  }
}

final reportesService = ReportesService();
